docs: add `x/reports` specs (#957)
desmos-labs/desmos: x/reports/spec/04-messages.md

---
id: messages
title: Messages
sidebar_label: Messages
slug: messages
---

# Msg Service

## Msg/CreateReport

> A report can be created using the `MsgCreateReport`, specifying the information you see below:

Reviewer: ```suggestion
A report can be created using the `MsgCreateReport`.
```

```js reference
https://github.com/desmos-labs/desmos/blob/020cf82788b667924d0f71f9d8f1fd87efa5b340/proto/desmos/reports/v1/msgs.proto#L34-L58
```

> The message is expected to fail if any of the following situations occurs:
> * The reporter does not have a profile;
> * The subspace does not exist;
> * One or more of the specified reasons IDs do not exist;
> * The reported does not have the permission to report content inside the subspace;
> * The report has already been made;
> * The report validation fails;
> * The report target type is not valid.

Reviewer: ```suggestion
It's expected to fail if:
* the reporter does not have a profile;
* the subspace does not exist;
* one of the specified reasons ids does not exist inside the subspace;
* the reported does not have the permission to report content within the subspace;
* another report for the same target has already been created by the same user;
* the report target does not exist.
```

## Msg/DeleteReport

> A report can be deleted using the `MsgDeleteReport`, specifying the information you see below:

Reviewer: ```suggestion
A report can be deleted using the `MsgDeleteReport`.
```

```js reference
https://github.com/desmos-labs/desmos/blob/020cf82788b667924d0f71f9d8f1fd87efa5b340/proto/desmos/reports/v1/msgs.proto#L77-L92
```

> The message is expected to fail if any of the following situations occurs:
> 
> * The subspace does not exist;
> * The report does not exist;
> * The signer does not have the permission to delete a report inside the subspace;
> * The signer is not the reporter;

Reviewer: ```suggestion
It's expected to fail if:
* the subspace does not exist;
* the report does not exist;
* the signer does not have the permission to delete a report within the subspace.
```

## Msg/SupportStandardReason

> The `MsgSupportStandardReason` can be used if you want to use the set of standard reasons specified in the module 
> params inside your dApp.

Reviewer: ```suggestion
A standard reason can be supported within a subspace using the `MsgSupportStandardReason`.
```

```js reference
https://github.com/desmos-labs/desmos/blob/020cf82788b667924d0f71f9d8f1fd87efa5b340/proto/desmos/reports/v1/msgs.proto#L99-L114
```

> The message is expected to fail if any of the following situations occurs:
> * The subspace does not exist;
> * The reason does not exist;
> * The signer does not have the permission to manage reasons inside the subspace;
> * The reason validation fails.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace does not exist;
* the reason does not exist;
* the signer does not have the permission to manage registered within inside the subspace.
```

## Msg/AddReason

> A reason can be added to your subspace using the `MsgAddReason`, specifying the information you see below:

Reviewer: ```suggestion
A reason can be added to a subspace using the `MsgAddReason`.
```

```js reference
https://github.com/desmos-labs/desmos/blob/020cf82788b667924d0f71f9d8f1fd87efa5b340/proto/desmos/reports/v1/msgs.proto#L128-L143
```

> The message is expected to fail if any of the following situations occurs:
> * The subspace reason does not exist;
> * The signer does not have the permission to manage reasons inside the subspace;
> * The system is unable to provide a valid next reason ID;
> * The reason validation fails.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace reason does not exist;
* the signer does not have the permission to manage reasons within the subspace;
* the reason name is either empty or blank.
```

## Msg/RemoveReason

> A previously added reason can be removed using the `MsgRemoveReason`, specifying the information you see below:

Reviewer: ```suggestion
A previously added reason can be removed using the `MsgRemoveReason`.
```

```js reference
https://github.com/desmos-labs/desmos/blob/020cf82788b667924d0f71f9d8f1fd87efa5b340/proto/desmos/reports/v1/msgs.proto#L156-L171
```

> The message is expected to fail if any of the following situations occurs:
> * The subspace reason does not exist;
> * The reason does not exist;
> * The signer does not have the permission to manage reasons inside the subspace.

Reviewer: ```suggestion
It's expected to fail if:
* the subspace reason does not exist;
* the reason does not exist;
* the signer does not have the permission to manage registered reasons within the subspace.
```
